src/algebra/category/Semigroup/basic.lean

```diff
@@ -0,0 +1,217 @@
+
+We introduce the bundled categories:
+* `Magma`
+* `AddMagma`
+* `Semigroup`
+* `AddSemigroup`
+along with the relevant forgetful functors between them.
+
+## TODO
+
+* Limits in these categories
+* adjoining a unit adjunction to Mon
+* free/forgetful adjunctions
+-/
+
+universes u v
+
+open category_theory
+
+/-- The category of magmas and magma morphisms. -/
+@[to_additive AddMagma]
+def Magma : Type (u+1) := bundled has_mul
+
+/-- The category of additive magmas and additive magma morphisms. -/
+add_decl_doc AddMagma
+
+namespace Magma
+
+@[to_additive]
+instance bundled_hom : bundled_hom @mul_hom :=
+⟨@mul_hom.to_fun, @mul_hom.id, @mul_hom.comp, @mul_hom.coe_inj⟩
+
+attribute [derive [has_coe_to_sort, large_category, concrete_category]] Magma AddMagma
+
+/-- Construct a bundled `Magma` from the underlying type and typeclass. -/
+@[to_additive]
+def of (M : Type u) [has_mul M] : Magma := bundled.of M
+
+/-- Construct a bundled `Magma` from the underlying type and typeclass. -/
+add_decl_doc AddMagma.of
+
+@[to_additive]
+instance : inhabited Magma := ⟨Magma.of empty⟩
+
+@[to_additive]
+instance (M : Magma) : has_mul M := M.str
+
+@[simp, to_additive] lemma coe_of (R : Type u) [has_mul R] : (Magma.of R : Type u) = R := rfl
+
+end Magma
+
+/-- The category of semigroups and semigroup morphisms. -/
+@[to_additive AddSemigroup]
+def Semigroup : Type (u+1) := bundled semigroup
+
+/-- The category of additive semigroups and semigroup morphisms. -/
+add_decl_doc AddSemigroup
+
+namespace Semigroup
+
+@[to_additive]
+instance : bundled_hom.parent_projection semigroup.to_has_mul := ⟨⟩
+
+attribute [derive [has_coe_to_sort, large_category, concrete_category]] Semigroup AddSemigroup
+
+/-- Construct a bundled `Semigroup` from the underlying type and typeclass. -/
+@[to_additive]
+def of (M : Type u) [semigroup M] : Semigroup := bundled.of M
+
+/-- Construct a bundled `AddSemigroup` from the underlying type and typeclass. -/
+add_decl_doc AddSemigroup.of
+
+@[to_additive]
+instance : inhabited Semigroup := ⟨Semigroup.of empty⟩
+
+@[to_additive]
+instance (M : Semigroup) : semigroup M := M.str
+
+@[simp, to_additive] lemma coe_of (R : Type u) [semigroup R] : (Semigroup.of R : Type u) = R := rfl
+
+@[to_additive has_forget_to_AddMagma]
+instance has_forget_to_Magma : has_forget₂ Semigroup Magma := bundled_hom.forget₂ _ _
+
+end Semigroup
+
+-- We verify that the coercions of morphisms to functions work correctly:
+example {R S : Magma}     (f : R ⟶ S) : (R : Type) → (S : Type) := f
+example {R S : Semigroup} (f : R ⟶ S) : (R : Type) → (S : Type) := f
+
+-- We verify that when constructing a morphism in `Semigroup`,
+-- when we construct the `to_fun` field, the types are presented as `↥R`,
+-- rather than `R.α` or (as we used to have) `↥(bundled.map semigroup.to_has_mul R)`.
+example (R : Semigroup.{u}) : R ⟶ R :=
+{ to_fun := λ X,
+  begin
+    match_target (R : Type u),
+    exact X
+  end ,
+  map_mul' := λ x y,
+  rfl, }
```

Comment:
I don't think we need to duplicate these examples across from `Mon`. If one breaks, surely the other will too.

Reply:
I removed them. Thanks.